File: HW12/HW12.py
import heapq
from typing import List

def minCostToSupplyWater(n: int, wells: List[int], pipes: List[List[int]]) -> int:
    edges = []
    well_costs = [0] + wells
    for i in range(n):
        edges.append([0, i + 1, wells[i]])
    for pipe in pipes:
        edges.append(pipe)
    edges.sort(key=lambda x: x[2])

    parent = list(range(n + 1))

    total_cost = 0
    for _, u, v, cost in edges:
        root_u = find(parent, u)
        root_v = find(parent, v)
        if root_u != root_v:
            parent[root_u] = root_v
            total_cost += cost

    return total_cost

def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]

File: HW12/HW12.java
import java.util.*;

public class Solution {
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        List<int[]> edges = new ArrayList<>();
        int[] wellCosts = new int[n + 1];
        for (int i = 0; i < n; i++) {
            wellCosts[i + 1] = wells[i];
            edges.add(new int[] {0, i + 1, wells[i]});
        }
        for (int[] pipe : pipes) {
            edges.add(new int[] {pipe[0], pipe[1], pipe[2]});
        }
        // Use Collections.sort() to sort the list of integer arrays
        Collections.sort(edges, (a, b) -> Integer.compare(a[2], b[2]));

        int[] parent = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            parent[i] = i;
        }

        int totalCost = 0;
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1], cost = edge[2];
            int rootU = find(parent, u);
            int rootV = find(parent, v);
            if (rootU != rootV) {
                parent[rootU] = rootV;
                totalCost += cost;
            }
        }

        return totalCost;
    }

    private int find(int[] parent, int x) {
        if (parent[x] != x) {
            parent[x] = find(parent, parent[x]);
        }
        return parent[x];
    }

    public static void main(String[] args) {
        int n = 2;
        int[] wells = {1, 1};
        int[][] pipes = {{1, 2, 1}, {1, 2, 2}};
        Solution solution = new Solution();
        int minCost = solution.minCostToSupplyWater(n, wells, pipes);
        System.out.println("Minimum cost to supply water: " + minCost);  // Expected output: 2
    }
}
